libs/constrained-generators/src/Constrained/Spec/Map.hs

```diff
@@ -308,7 +350,537 @@ dom_ ::
+    fixedKrandomV ks n usedvs usedks k = do
+      v <-
+        genFromSpecT @fn @_ @m
+          ( constrained $ \v ->
+              [ assert $ not_ (elem_ v (lit usedvs))
+              , satisfies (pair_ (lit k) v) ok
+              ]
+          )
+      loopK ((k, v) : ans) ks n dups (v : usedvs) (k : usedks) ok
+
+-- | Generate, or find a match , k, (in 'mustks') for v, that satisfies ok (k,v)
+genOrFindMatchFor ::
+  (MonadGenError m, HasSpec fn v, HasSpec fn k, Ord k) =>
+  v ->
+  Set k ->
+  Set k ->
+  Specification fn (k, v) ->
+  GenT m k
+genOrFindMatchFor v mustks usedks ok =
+  explain
+    [ "Find a k-match for v=" ++ show v ++ " where k in " ++ show (Set.toList mustks)
+    , "  or"
+    , "Gen a k-match for v="
+        ++ show v
+        ++ " where k not in "
+        ++ show (Set.toList (Set.union mustks usedks))
+    ]
+    (trySpecs [findMatch, genMatch])
+  where
+    genMatch =
+      constrained $ \key ->
+        [ satisfies (pair_ key (lit v)) ok
+        , assert $ not_ (member_ key (lit (Set.union mustks usedks)))
+        ]
+    findMatch =
+      constrained $ \key ->
+        [ satisfies (pair_ key (lit v)) ok
+        , assert $ member_ key (lit mustks)
+        ]
+
+-- =================================================================================
+-- More accurate than the Arbitrary instance
+-- =================================================================================
+
+-- | Usefull for generating (MapSpec fn Integer Integer) for use in tests
+--   Most of the MapSpec's generated should be valid, with a good distribution
+--   of different sizes for mustVals and mustKeys, and a high probability of
+--   satisfying kvSpec (key >= val)
+genMapSpec :: forall fn. BaseUniverse fn => Gen (MapSpec fn Integer Integer)
+genMapSpec = do
+  (mustKeyLen, mustValLen, _slack) <- genSizes
+  mapsize <- genSizeSpec @fn (mustKeyLen + mustValLen) 6
+  let biggest = maxSize mapsize
+  let ok x = x <= biggest && x > mustValLen
+  let genOk = if mustValLen == biggest then pure biggest else arbitrary `suchThat` ok -- ((getSmall <$> arbitrary) `suchThat` ok)
+  valsize <-
+    frequency
+      [ (1, pure $ Just (MinValSize (typeSpec (NumSpecInterval (Just mustValLen) (Just biggest)))))
+      , (1, pure $ Just (MinValSize (typeSpec (NumSpecInterval Nothing (Just biggest)))))
+      , (1, pure Nothing)
+      , (1, Just . MinValSize . MemberSpec . (mustValLen :) <$> vectorOf 5 genOk)
+      ]
+  mustval <- vectorOf (fromInteger mustValLen) arbitrary
+  let avg =
+        if null mustval
+          then 0 :: Integer
+          else sum mustval `div` toInteger (length mustval)
+      goodMust = constrained (\n -> n >=. (lit avg))
+  mustkey <- genFromGenT $ (Set.fromList <$> genDistinctList @_ @fn goodMust mustKeyLen [])
+  let kvSpec =
+        constrained
+          ( \p ->
+              match p (\key val -> [assert $ key >=. val])
+          )
+  pure (MapSpec Nothing mustkey mustval mapsize kvSpec NoFold valsize)
```

Comment:
I would be good with replacing the current arbitrary instance with this if we can also generate the old garbage map instance ~1/3 of the time to make sure we cover all the really nasty cases.

Reply:
We would have to generalize it from (Map Integer Integer) to (Map k v), lets do this in a follow on PR.